File: backend/src/main/java/com/leadgrowth/service/LeadQueueService.java
package com.leadgrowth.service;

import com.leadgrowth.dto.LeadDto;
import com.leadgrowth.entity.AssignmentLog;
import com.leadgrowth.entity.Lead;
import com.leadgrowth.entity.LeadAssignment;
import com.leadgrowth.entity.User;
import com.leadgrowth.repository.*;
import com.leadgrowth.websocket.WebSocketManager;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.*;
import java.util.stream.Collectors;

@Service
public class LeadQueueService {

    private final LeadRepository leadRepository;
    private final UserRepository userRepository;
    private final LeadAssignmentRepository leadAssignmentRepository;
    private final AssignmentLogRepository assignmentLogRepository;
    private final NotificationRepository notificationRepository;
    private final WebSocketManager webSocketManager;
    private final AuditService auditService;

    public LeadQueueService(
            LeadRepository leadRepository,
            UserRepository userRepository,
            LeadAssignmentRepository leadAssignmentRepository,
            AssignmentLogRepository assignmentLogRepository,
            NotificationRepository notificationRepository,
            WebSocketManager webSocketManager,
            AuditService auditService
    ) {
        this.leadRepository = leadRepository;
        this.userRepository = userRepository;
        this.leadAssignmentRepository = leadAssignmentRepository;
        this.assignmentLogRepository = assignmentLogRepository;
        this.notificationRepository = notificationRepository;
        this.webSocketManager = webSocketManager;
        this.auditService = auditService;
    }

    public List<LeadDto> getUnassignedLeadQueue(Long workspaceId) {
        List<Lead> leads = leadRepository.findByWorkspaceIdOrderByCreatedAtDesc(workspaceId);
        return leads.stream()
                .filter(l -> l.getAssignedTo() == null || "IN_QUEUE".equals(l.getQueueStatus()))
                .map(this::convertToDto)
                .collect(Collectors.toList());
    }

    public List<LeadDto> bulkAssignLeads(List<Long> leadIds, Long targetUserId, String actorEmail) {
        User actor = userRepository.findByEmail(actorEmail)
                .orElseThrow(() -> new RuntimeException("Actor not found"));
        User targetUser = userRepository.findById(targetUserId)
                .orElseThrow(() -> new RuntimeException("Target user not found"));

        List<LeadDto> assignedList = new ArrayList<>();
        for (Long id : leadIds) {
            Lead lead = leadRepository.findById(id).orElse(null);
            if (lead != null && lead.getWorkspace().getId().equals(actor.getWorkspace().getId())) {
                lead.setAssignedTo(targetUser);
                lead.setQueueStatus("ASSIGNED");
                Lead saved = leadRepository.save(lead);

                auditService.logAction(lead.getWorkspace(), actor, "LEAD_ASSIGNED", "LEAD", saved.getId(),
                        "Bulk assigned lead " + saved.getName() + " to " + targetUser.getFullName());

                assignedList.add(convertToDto(saved));
            }
        }
        return assignedList;
    }

    public LeadDto autoAssignLead(Long leadId, String actorEmail) {
        User actor = userRepository.findByEmail(actorEmail)
                .orElseThrow(() -> new RuntimeException("Actor user not found"));

        Lead lead = leadRepository.findById(leadId)
                .orElseThrow(() -> new RuntimeException("Lead not found"));

        Long workspaceId = actor.getWorkspace().getId();
        User bestAssignee = findBestAssignee(workspaceId);

        if (bestAssignee != null) {
            lead.setAssignedTo(bestAssignee);
            lead.setQueueStatus("ASSIGNED");
            Lead saved = leadRepository.save(lead);

            bestAssignee.setLastAssignedAt(LocalDateTime.now());
            userRepository.save(bestAssignee);

            leadAssignmentRepository.save(new LeadAssignment(saved, bestAssignee));
            assignmentLogRepository.save(new AssignmentLog(
                    actor.getWorkspace(), "LEAD", saved.getId(), bestAssignee, "Auto-Assigned via Hybrid Algorithm (Availability & Workload)"
            ));

            auditService.logAction(actor.getWorkspace(), actor, "LEAD_AUTO_ASSIGNED", "LEAD", saved.getId(),
                    "Auto-assigned lead '" + saved.getName() + "' to " + bestAssignee.getFullName());

            return convertToDto(saved);
        } else {
            throw new IllegalStateException("No available team members currently eligible for auto-assignment.");
        }
    }

    public LeadDto triggerIdlePreventionSweep(String userEmail) {
        User user = userRepository.findByEmail(userEmail)
                .orElseThrow(() -> new RuntimeException("User not found"));

        Long workspaceId = user.getWorkspace().getId();
        List<Lead> queue = leadRepository.findByWorkspaceIdOrderByCreatedAtDesc(workspaceId).stream()
                .filter(l -> l.getAssignedTo() == null || "IN_QUEUE".equals(l.getQueueStatus()))
                .collect(Collectors.toList());

        if (queue.isEmpty()) {
            return null;
        }

        // Draw top priority lead from queue
        Lead leadToAssign = queue.get(0);
        leadToAssign.setAssignedTo(user);
        leadToAssign.setQueueStatus("ASSIGNED");
        Lead saved = leadRepository.save(leadToAssign);

        user.setLastAssignedAt(LocalDateTime.now());
        userRepository.save(user);

        leadAssignmentRepository.save(new LeadAssignment(saved, user));
        assignmentLogRepository.save(new AssignmentLog(
                user.getWorkspace(), "LEAD", saved.getId(), user, "Assigned via User Idle Prevention Sweep"
        ));

        return convertToDto(saved);
    }

    private User findBestAssignee(Long workspaceId) {
        List<User> candidates = userRepository.findByWorkspaceId(workspaceId).stream()
                .filter(u -> "ACTIVE".equalsIgnoreCase(u.getStatus()))
                .filter(u -> !"OFFLINE".equals(u.getAvailabilityStatus()) && !"ON_LEAVE".equals(u.getAvailabilityStatus()))
                .collect(Collectors.toList());

        if (candidates.isEmpty()) return null;

        // Sort by availability (AVAILABLE > ON_BREAK > BUSY) then least recently assigned
        candidates.sort((u1, u2) -> {
            int score1 = getAvailabilityScore(u1.getAvailabilityStatus());
            int score2 = getAvailabilityScore(u2.getAvailabilityStatus());
            if (score1 != score2) return Integer.compare(score2, score1);

            LocalDateTime t1 = u1.getLastAssignedAt() != null ? u1.getLastAssignedAt() : LocalDateTime.MIN;
            LocalDateTime t2 = u2.getLastAssignedAt() != null ? u2.getLastAssignedAt() : LocalDateTime.MIN;
            return t1.compareTo(t2);
        });

        return candidates.get(0);
    }

    private int getAvailabilityScore(String status) {
        if (status == null) return 0;
        switch (status.toUpperCase()) {
            case "AVAILABLE": return 3;
            case "ON_BREAK": return 2;
            case "BUSY": return 1;
            default: return 0;
        }
    }

    private LeadDto convertToDto(Lead lead) {
        return LeadDto.builder()
                .id(lead.getId())
                .name(lead.getName())
                .email(lead.getEmail())
                .phone(lead.getPhone())
                .sourcePlatform(lead.getSourcePlatform())
                .campaignName(lead.getCampaignName())
                .campaignId(lead.getCampaign() != null ? lead.getCampaign().getId() : null)
                .status(lead.getStatus())
                .assignedToId(lead.getAssignedTo() != null ? lead.getAssignedTo().getId() : null)
                .assignedToName(lead.getAssignedTo() != null ? lead.getAssignedTo().getFullName() : "Unassigned")
                .qualityScore(lead.getQualityScore() != null ? lead.getQualityScore() : 75)
                .qualityTier(lead.getQualityTier() != null ? lead.getQualityTier() : "WARM")
                .conversionProbability(lead.getConversionProbability() != null ? lead.getConversionProbability() : 75.0)
                .queueStatus(lead.getQueueStatus() != null ? lead.getQueueStatus() : (lead.getAssignedTo() != null ? "ASSIGNED" : "IN_QUEUE"))
                .createdAt(lead.getCreatedAt())
                .build();
    }
}

File: backend-dotnet/Services/LeadQueueService.cs
using LeadGrowth.Data;
using LeadGrowth.DTOs;
using LeadGrowth.Models;
using Microsoft.EntityFrameworkCore;

namespace LeadGrowth.Services;

public class LeadQueueService : ILeadQueueService
{
    private readonly LeadGrowthDbContext _context;
    private readonly IWebSocketManagerService _webSocketService;

    public LeadQueueService(LeadGrowthDbContext context, IWebSocketManagerService webSocketService)
    {
        _context = context;
        _webSocketService = webSocketService;
    }

    public async Task<List<LeadDto>> GetUnassignedLeadQueueAsync(long workspaceId)
    {
        var leads = await _context.Leads
            .Include(l => l.AssignedTo)
            .Include(l => l.Campaign)
            .Where(l => l.WorkspaceId == workspaceId)
            .OrderByDescending(l => l.CreatedAt)
            .ToListAsync();

        return leads
            .Where(l => l.AssignedToId == null || string.Equals("IN_QUEUE", l.QueueStatus, StringComparison.OrdinalIgnoreCase))
            .Select(ConvertToDto)
            .ToList();
    }

    public async Task<List<LeadDto>> BulkAssignLeadsAsync(List<long> leadIds, long targetUserId, string actorEmail)
    {
        var email = actorEmail.Trim().ToLower();
        var actor = await _context.Users.FirstOrDefaultAsync(u => u.Email == email);
        if (actor == null)
        {
            throw new KeyNotFoundException("Actor not found");
        }

        var targetUser = await _context.Users.FirstOrDefaultAsync(u => u.Id == targetUserId);
        if (targetUser == null)
        {
            throw new KeyNotFoundException("Target user not found");
        }

        var assignedList = new List<LeadDto>();
        foreach (var id in leadIds)
        {
            var lead = await _context.Leads
                .Include(l => l.Workspace)
                .Include(l => l.AssignedTo)
                .Include(l => l.Campaign)
                .FirstOrDefaultAsync(l => l.Id == id);

            if (lead != null && lead.WorkspaceId == actor.WorkspaceId)
            {
                lead.AssignedToId = targetUser.Id;
                lead.ProgressPercentage = LeadService.CalculateProgressPercentage(lead.Status, true);
                lead.AssignedTo = targetUser;
                lead.AssignedById = actor.Id;
                lead.AssignedDate = DateTime.UtcNow;
                lead.QueueStatus = "ASSIGNED";
                await _context.SaveChangesAsync();

                var log = new AuditLog
                {
                    WorkspaceId = lead.WorkspaceId,
                    UserId = actor.Id,
                    Action = "LEAD_ASSIGNED",
                    TargetType = "LEAD",
                    TargetId = lead.Id,
                    Description = $"Bulk assigned lead {lead.Name} to {targetUser.FullName}",
                    CreatedAt = DateTime.UtcNow
                };
                _context.AuditLogs.Add(log);

                var notif = new Notification
                {
                    UserId = targetUser.Id,
                    Title = "New Lead Assigned",
                    Message = $"You have been assigned lead '{lead.Name}' by {actor.FullName}.",
                    IsRead = false,
                    CreatedAt = DateTime.UtcNow
                };
                _context.Notifications.Add(notif);
                await _context.SaveChangesAsync();

                var dto = ConvertToDto(lead);
                assignedList.Add(dto);

                try
                {
                    await _webSocketService.BroadcastNotificationAsync(targetUser.Id, new
                    {
                        id = notif.Id,
                        title = notif.Title,
                        message = notif.Message,
                        leadId = lead.Id,
                        createdAt = notif.CreatedAt,
                        type = "LEAD"
                    });
                    if (lead.WorkspaceId > 0)
                    {
                        await _webSocketService.BroadcastLeadAsync(lead.WorkspaceId, dto);
                    }
                }
                catch {}
            }
        }

        if (actor.WorkspaceId.HasValue)
        {
            try
            {
                await _webSocketService.BroadcastWorkspaceNotificationAsync(actor.WorkspaceId.Value, new
                {
                    type = "QUEUE_LEADS_ASSIGNED",
                    count = assignedList.Count,
                    targetUserId = targetUser.Id,
                    targetUserName = targetUser.FullName,
                    assignedByName = actor.FullName,
                    message = $"{actor.FullName} assigned {assignedList.Count} lead(s) to {targetUser.FullName}."
                });
            }
            catch {}
        }

        return assignedList;
    }

    public async Task<LeadDto> AutoAssignLeadAsync(long leadId, string actorEmail)
    {
        var email = actorEmail.Trim().ToLower();
        var actor = await _context.Users
            .Include(u => u.Workspace)
            .FirstOrDefaultAsync(u => u.Email == email);

        if (actor == null)
        {
            throw new KeyNotFoundException("Actor user not found");
        }

        var lead = await _context.Leads
            .Include(l => l.Workspace)
            .Include(l => l.AssignedTo)
            .Include(l => l.Campaign)
            .FirstOrDefaultAsync(l => l.Id == leadId);

        if (lead == null)
        {
            throw new KeyNotFoundException("Lead not found");
        }

        var bestAssignee = await FindBestAssigneeAsync(actor.WorkspaceId!.Value);

        if (bestAssignee != null)
        {
            lead.AssignedToId = bestAssignee.Id;
            lead.ProgressPercentage = LeadService.CalculateProgressPercentage(lead.Status, true);
            lead.AssignedTo = bestAssignee;
            lead.AssignedById = actor.Id;
            lead.AssignedDate = DateTime.UtcNow;
            lead.QueueStatus = "ASSIGNED";
            await _context.SaveChangesAsync();

            bestAssignee.LastAssignedAt = DateTime.UtcNow;
            await _context.SaveChangesAsync();

            var assignment = new LeadAssignment
            {
                LeadId = lead.Id,
                UserId = bestAssignee.Id,
                AssignedAt = DateTime.UtcNow
            };
            _context.LeadAssignments.Add(assignment);

            var assignLog = new AssignmentLog
            {
                WorkspaceId = actor.WorkspaceId.Value,
                LeadId = lead.Id,
                UserId = bestAssignee.Id,
                Strategy = "Auto-Assigned via Hybrid Algorithm (Availability & Workload)",
                AssignedAt = DateTime.UtcNow
            };
            _context.AssignmentLogs.Add(assignLog);

            var auditLog = new AuditLog
            {
                WorkspaceId = actor.WorkspaceId.Value,
                UserId = actor.Id,
                Action = "LEAD_AUTO_ASSIGNED",
                TargetType = "LEAD",
                TargetId = lead.Id,
                Description = $"Auto-assigned lead '{lead.Name}' to {bestAssignee.FullName}",
                CreatedAt = DateTime.UtcNow
            };
            _context.AuditLogs.Add(auditLog);

            var notif = new Notification
            {
                UserId = bestAssignee.Id,
                Title = "New Lead Assigned",
                Message = $"You have been auto-assigned lead '{lead.Name}'.",
                IsRead = false,
                CreatedAt = DateTime.UtcNow
            };
            _context.Notifications.Add(notif);
            await _context.SaveChangesAsync();

            var dto = ConvertToDto(lead);
            try
            {
                await _webSocketService.BroadcastNotificationAsync(bestAssignee.Id, new
                {
                    id = notif.Id,
                    title = notif.Title,
                    message = notif.Message,
                    leadId = lead.Id,
                    createdAt = notif.CreatedAt,
                    type = "LEAD"
                });
                if (lead.WorkspaceId > 0)
                {
                    await _webSocketService.BroadcastLeadAsync(lead.WorkspaceId, dto);
                }
            }
            catch {}

            return dto;
        }
        else
        {
            throw new InvalidOperationException("No available team members currently eligible for auto-assignment.");
        }
    }

    public async Task<LeadDto?> TriggerIdlePreventionSweepAsync(string userEmail)
    {
        var email = userEmail.Trim().ToLower();
        var user = await _context.Users
            .Include(u => u.Workspace)
            .FirstOrDefaultAsync(u => u.Email == email);

        if (user == null)
        {
            throw new KeyNotFoundException("User not found");
        }

        var queue = await _context.Leads
            .Include(l => l.AssignedTo)
            .Include(l => l.Campaign)
            .Where(l => l.AssignedToId == null || string.Equals("IN_QUEUE", l.QueueStatus))
            .OrderByDescending(l => l.CreatedAt)
            .ToListAsync();

        if (queue.Count == 0)
        {
            return null;
        }

        var leadToAssign = queue[0];
        leadToAssign.AssignedToId = user.Id;
        leadToAssign.AssignedTo = user;
        leadToAssign.QueueStatus = "ASSIGNED";
        if (user.WorkspaceId.HasValue)
        {
            leadToAssign.WorkspaceId = user.WorkspaceId.Value;
        }

        await _context.SaveChangesAsync();

        user.LastAssignedAt = DateTime.UtcNow;
        await _context.SaveChangesAsync();

        var assignment = new LeadAssignment
        {
            LeadId = leadToAssign.Id,
            UserId = user.Id,
            AssignedAt = DateTime.UtcNow
        };
        _context.LeadAssignments.Add(assignment);

        if (user.WorkspaceId.HasValue)
        {
            var assignLog = new AssignmentLog
            {
                WorkspaceId = user.WorkspaceId.Value,
                LeadId = leadToAssign.Id,
                UserId = user.Id,
                Strategy = "Assigned via User Idle Prevention Sweep",
                AssignedAt = DateTime.UtcNow
            };
            _context.AssignmentLogs.Add(assignLog);
        }

        var notif = new Notification
        {
            UserId = user.Id,
            Title = "Lead Assigned via Idle Sweep",
            Message = $"You picked up queue lead '{leadToAssign.Name}' via Idle Prevention Sweep.",
            IsRead = false,
            CreatedAt = DateTime.UtcNow
        };
        _context.Notifications.Add(notif);
        await _context.SaveChangesAsync();

        var dto = ConvertToDto(leadToAssign);
        try
        {
            await _webSocketService.BroadcastNotificationAsync(user.Id, new
            {
                id = notif.Id,
                title = notif.Title,
                message = notif.Message,
                leadId = leadToAssign.Id,
                createdAt = notif.CreatedAt,
                type = "LEAD"
            });
            if (leadToAssign.WorkspaceId > 0)
            {
                await _webSocketService.BroadcastLeadAsync(leadToAssign.WorkspaceId, dto);
            }
        }
        catch {}

        return dto;
    }

    private async Task<User?> FindBestAssigneeAsync(long workspaceId)
    {
        var candidates = await _context.Users
            .Where(u => u.WorkspaceId == workspaceId && !string.Equals("SUSPENDED", u.Status))
            .Where(u => !string.Equals("OFFLINE", u.AvailabilityStatus) && !string.Equals("ON_LEAVE", u.AvailabilityStatus))
            .ToListAsync();

        if (candidates.Count == 0) return null;

        var sorted = candidates
            .OrderByDescending(u => GetAvailabilityScore(u.AvailabilityStatus))
            .ThenBy(u => u.LastAssignedAt ?? DateTime.MinValue)
            .ToList();

        return sorted[0];
    }

    private static int GetAvailabilityScore(string? status)
    {
        if (string.IsNullOrEmpty(status)) return 0;
        return status.ToUpper() switch
        {
            "AVAILABLE" => 3,
            "ON_BREAK" => 2,
            "BUSY" => 1,
            _ => 0
        };
    }

    private static LeadDto ConvertToDto(Lead lead)
    {
        return new LeadDto
        {
            Id = lead.Id,
            Name = lead.Name,
            Email = lead.Email,
            Phone = lead.Phone,
            SourcePlatform = lead.SourcePlatform,
            CampaignName = lead.CampaignName,
            CampaignId = lead.CampaignId,
            Status = lead.Status,
            AssignedToId = lead.AssignedToId,
            AssignedToName = lead.AssignedTo != null ? lead.AssignedTo.FullName : "Unassigned",
            QualityScore = lead.QualityScore ?? 75,
            QualityTier = lead.QualityTier ?? "WARM",
            ConversionProbability = lead.ConversionProbability ?? 75.0,
            QueueStatus = lead.QueueStatus ?? (lead.AssignedToId.HasValue ? "ASSIGNED" : "IN_QUEUE"),
            CreatedAt = lead.CreatedAt
        };
    }
}
